### PR TITLE
Early-stop and more features

### PR DESCRIPTION
**Pull request by unmonoqueteclea**
Added early-stop for executions with bad performance.
Added filtered dataset at the output.
Used phi - standard deviation as performance metric

### CODE REVIEW

File: best-first-cv/metadata.json [hunk 1]
```diff
@@ -38,13 +38,30 @@
...
+      "name": "early-stop-performance",
+      "type": "number",
+      "default": 1,
+      "description": "Performance improvement value (in %) used as a threshold to consider good performance of a feature"
+    },
+    {
+      "name": "max-bad-percentage",
+      "type": "number",
+      "default": 20,
+      "description": "(%) Maximum number of iterations with bad performance, as percentage of the number of initial features"
```
Comment (jaor) [inline best-first-cv/metadata.json]: nit: you already say it's a percentage, so the `(%)` prefix is kind of redundant... if you still like it, then it should be used too in the previous input's description, for consistency.

File: best-first-cv/script.whizzml [hunk 2]
```diff
@@ -408,55 +448,127 @@
...
                          objective-id
                          options
                          k-folds
-                         pre-selected)
+                         pre-selected
+                         early-stop-performance
+                         max-bad-pecentage)
   (let (obj-id (get-objective dataset-id objective-id)
-        k-fold-datasets (create-k-folds dataset-id k-folds)
+	k-fold-datasets (create-k-folds dataset-id k-folds)
```
Comment (jaor) [inline best-first-cv/script.whizzml:486]: this should be aligned with obj-id

File: best-first-cv/script.whizzml [hunk 3]
```diff
@@ -408,55 +448,127 @@
...
         fields (resource-fields dataset-id)
+	max-bad (* (/ max-bad-percentage 100) (count fields))
+	max-bad (if (< max-bad 1)
+		  1
+		  max-bad)
```
Comment (jaor) [inline best-first-cv/script.whizzml]: `(max 1 max-bad)` is nicer

File: best-first-cv/script.whizzml [hunk 4]
```diff
@@ -408,55 +448,127 @@
...
+;;
+;; Inputs:
+;;   queue: (list) List of maps with results from previous iterations
+;;   next-feat-info: (map) Results from last selected feature
+;;   early-stop-performance: (number) Minimum required performance improvement
+;;   bad-count: (number) Current number of features with bad performance
+;;
+;; Output: (number) Updated number of features with bad performance
+;;
+(define(check-bad-performance queue
```
Comment (jaor) [inline best-first-cv/script.whizzml]: missing whitespace `(define (check`

File: best-first-cv/script.whizzml [hunk 5]
```diff
@@ -468,19 +580,39 @@
...
+;;    bad-count: (number) Current number of features with bad performance
+;;   
+;;
+;; Output: (map) Response with selected fields names and iterations info
+;;
+(define (response dataset-id k-fold-datasets selected queue bad-count)
+ ;; We have to remove fields with bad performance 
+ (let  (selected (if (> bad-count 0)
+                     (reverse (drop  bad-count selected))
+                     (reverse selected )))
```
Comment (jaor) [inline best-first-cv/script.whizzml]: extra whitespace after `selected`

File: best-first-cv/script.whizzml [hunk 6]
```diff
@@ -528,15 +675,36 @@
...
-        vs cvs
+        ;; Obtains phi-stdev from each cross validation
+        vs (map (lambda (ev) (ev ["phi-stdev"])) cvs)
         value-map (make-map potentials vs)
         max-val (get-max vs)
-        choose-best (lambda (id) (and (= max-val (value-map id)) id)))
+        choose-best (lambda (id) (and (= max-val (value-map id)) id))
+        phi (some (lambda (n) (if (= (n ["phi-stdev"]) max-val)
+                                  (n ["phi"]) 
+                                  false))
```
Comment (jaor) [inline best-first-cv/script.whizzml:683]: `if`s returning booleans are almost always clunky.  better: `(and (= (max-val (n "phi-stdev")) (n "phi"))` or `(when (= max-val (n "phi-stdev")) (n "phi"))`

**Review by jaor — CHANGES_REQUESTED** [review]
looks mostly good, just a couple of style nits

File: best-first-cv/script.whizzml (hunk 2 above)
Comment (unmonoqueteclea) [reply best-first-cv/script.whizzml:486]: It is strange, in the actual file, the indentation seems correct. Please, check new commit.

**Review by jaor — APPROVED** [review]

**Comment by jaor** [comment]
looks good, merged!

**Pull request merged**